kv/kvserver: TestStoreRangeMergeTimestampCacheCausality failed

kv/kvserver.TestStoreRangeMergeTimestampCacheCausality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9136920?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9136920?buildTab=artifacts#/) on master @ [53dbb86acb1d48309530181b94838faf937084d3](https://github.com/cockroachdb/cockroach/commits/53dbb86acb1d48309530181b94838faf937084d3):


```
=== RUN   TestStoreRangeMergeTimestampCacheCausality
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality855634793
    test_log_scope.go:79: use -show-logs to present logs inline
    client_merge_test.go:1030: [NotLeaseHolderError] stale proposal: command was proposed under lease #4 but is being applied under lease: repl=(n2,s2):2 seq=5 start=1679207140.779003082,0 epo=1 pro=1679207140.779553612,0; r61: replica not lease holder; current lease is repl=(n2,s2):2 seq=5 start=1679207140.779003082,0 epo=1 pro=1679207140.779553612,0
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality855634793
--- FAIL: TestStoreRangeMergeTimestampCacheCausality (2.17s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeTimestampCacheCausality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-25626

kv/kvserver.TestStoreRangeMergeTimestampCacheCausality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9136920?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/9136920?buildTab=artifacts#/) on master @ [53dbb86acb1d48309530181b94838faf937084d3](https://github.com/cockroachdb/cockroach/commits/53dbb86acb1d48309530181b94838faf937084d3):


```
=== RUN   TestStoreRangeMergeTimestampCacheCausality
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality3661691425
    test_log_scope.go:79: use -show-logs to present logs inline
    client_merge_test.go:1030: [NotLeaseHolderError] stale proposal: command was proposed under lease #4 but is being applied under lease: repl=(n2,s2):2 seq=5 start=1679207442.874056053,0 epo=1 pro=1679207442.874765213,0; r61: replica not lease holder; current lease is repl=(n2,s2):2 seq=5 start=1679207442.874056053,0 epo=1 pro=1679207442.874765213,0
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality3661691425
--- FAIL: TestStoreRangeMergeTimestampCacheCausality (1.81s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeTimestampCacheCausality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeMergeTimestampCacheCausality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9132676?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9132676?buildTab=artifacts#/) on master @ [53dbb86acb1d48309530181b94838faf937084d3](https://github.com/cockroachdb/cockroach/commits/53dbb86acb1d48309530181b94838faf937084d3):


```
=== RUN   TestStoreRangeMergeTimestampCacheCausality
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality4167804785
    test_log_scope.go:79: use -show-logs to present logs inline
    client_merge_test.go:1030: [NotLeaseHolderError] stale proposal: command was proposed under lease #4 but is being applied under lease: repl=(n2,s2):2 seq=5 start=1679212986.599850280,0 epo=1 pro=1679212986.600480768,0; r61: replica not lease holder; current lease is repl=(n2,s2):2 seq=5 start=1679212986.599850280,0 epo=1 pro=1679212986.600480768,0
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality4167804785
--- FAIL: TestStoreRangeMergeTimestampCacheCausality (1.78s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeTimestampCacheCausality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeMergeTimestampCacheCausality [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9132680?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9132680?buildTab=artifacts#/) on master @ [53dbb86acb1d48309530181b94838faf937084d3](https://github.com/cockroachdb/cockroach/commits/53dbb86acb1d48309530181b94838faf937084d3):


```
=== RUN   TestStoreRangeMergeTimestampCacheCausality
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality977970002
    test_log_scope.go:79: use -show-logs to present logs inline
    client_merge_test.go:1030: [NotLeaseHolderError] stale proposal: command was proposed under lease #4 but is being applied under lease: repl=(n2,s2):2 seq=5 start=1679212924.467215947,0 epo=1 pro=1679212924.471439359,0; r61: replica not lease holder; current lease is repl=(n2,s2):2 seq=5 start=1679212924.467215947,0 epo=1 pro=1679212924.471439359,0
    panic.go:540: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeTimestampCacheCausality977970002
--- FAIL: TestStoreRangeMergeTimestampCacheCausality (9.66s)
```
<p>Parameters: <code>TAGS=bazel,gss,race</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeTimestampCacheCausality.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -1017,6 +1017,11 @@ func TestStoreRangeMergeTimestampCacheCausality(t *testing.T) {
 		if !lhsRepl1.OwnsValidLease(ctx, tc.Servers[1].Clock().NowAsClockTimestamp()) {
 			return errors.New("s2 does not own valid lease for lhs range")
 		}
+		if !kvserver.ExpirationLeasesOnly.Get(&tc.Server(0).ClusterSettings().SV) { // metamorphic
+			if lhsRepl1.CurrentLeaseStatus(ctx).Lease.Type() != roachpb.LeaseEpoch {
+				return errors.Errorf("lease still an expiration based lease")
+			}
+		}
 		return nil
 	})
 

